Extensions/v0.7 extensions (#86)
kwilteam/docs: docs/extensions/cross-chain-data/oracles.mdx

> ---

Reviewer: Can we change this file name to be "overview.mdx"?

sidebar_position: 1
sidebar_label: "Overview"
id: oracles-overview

> title: Oracles

Reviewer: ```suggestion
title: Event-driven Oracles Overview
```

description: Build custom event-driven oracles.
slug: /extensions/oracles-overview
---

> Kwil networks have the ability to come to consensus on external system events, and trigger subsequent state changes based on the result of those events. This allows for network changes to be automatically triggered by external systems, such as other blockchain networks.

Reviewer: ```suggestion
Kwil networks have the ability to come to consensus on external system events and trigger subsequent state changes based on the result of those events. This allows for network changes to be automatically triggered by external systems, such as other blockchain networks.
```

> Example usecases for Kwil oracles include:

Reviewer: ```suggestion
Example use cases for Kwil oracles include:
```

- Giving users native network gas when they burn an ERC20 token on Ethereum.
- Updating a validator's power when they stake an SPL token on Solana.
- Automatically syncing data from another Kwil network.

## Anatomy of an Oracle

> Oracles are made up of two distinct components: event listeners, and resolutions.

Reviewer: I think the docs should be consistent on which order "event listeners" and "resolutions" are presented. Everywhere else, resolutions are presented first. But here, it logically makes sense to present event listeners first (because nodes must agree on events before voting on the resolution). Should we change the rest of the docs to present event listeners first?

> [Event Listeners](/docs/extensions/cross-chain-data/listeners.mdx) are responsible for observing external systems and storing information about the events that occur. When an event listener detects an event, it stores the event in its local database. The node will then broadcast information about the event to the rest of the network, until the event is recognized by a majority of the network.

Reviewer: ```suggestion
[**Event Listeners**](/docs/extensions/cross-chain-data/listeners.mdx) are responsible for observing external systems and storing information about the events that occur. When an event listener detects an event, it stores the event in its local database. The node will then broadcast information about the event to the rest of the network, until the event is recognized by a majority of the network.
```

Reviewer: Also, does the event store clear after the nodes agree?

> [Resolutions](/docs/extensions/cross-chain-data/resolutions.mdx) are responsible for taking the information about the event from the event listener, and using it to trigger a state change in the network. When a resolution is triggered, it will broadcast information about the state change to the rest of the network, until the state change is recognized by a majority of the network.

Reviewer: ```suggestion
[**Resolutions**](/docs/extensions/cross-chain-data/resolutions.mdx) are responsible for taking the information about the event from the event listener, and using it to trigger a state change in the network. When a resolution is triggered, it will broadcast information about the state change to the rest of the network, until the state change is recognized by a majority of the network.
```

## Event Lifecycle

The process by which the network agrees on external events can be broken down into the following steps:

> 1. Nodes observe an event in an external system, and broadcast this information to the rest of the network. The hash of the observed event, with the node it was observed by, is mined into a block.

Reviewer: ```suggestion
1. Nodes first observe an event in an external system, then broadcast this information to the rest of the network. The hash of the observed event, with the node it was observed by, is mined into a block.
```

2. A node who has seen the event has become a block proposer. They will include the full event data in the proposed block.
3. Once enough nodes have acknowledged the event, and the full event data has been included, the corresponding changes will be made to the network state.

### Step 1: Observing an Event

> When an event is observed by a node, the node will store the event in its local database. The node will immediately broadcast the ID of the event to the rest of the network.

Reviewer: ```suggestion
When a node observes an event, the node will store the event in its local database. The node will immediately broadcast the ID of the event to the rest of the network.
```
